pkg/export/secrets/manager.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	failedSecretConfigs = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_kubernetes_failed_secret_configs",
+			Help: "Current number of secret configurations that failed to load.",
+		},
+	)
+	secretsTotal = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_kubernetes_secrets_total",
+			Help: "Current number of secrets.",
+		},
+	)
+)
+
+func yamlSerialize(obj any) ([]byte, error) {
+	if obj == nil {
+		return []byte{}, nil
+	}
+	var buf bytes.Buffer
+	encoder := yaml.NewEncoder(&buf)
+	if err := encoder.Encode(obj); err != nil {
+		return nil, err
+	}
+	if err := encoder.Close(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func yamlEqual(x, y any) (bool, error) {
+	yamlX, err := yamlSerialize(x)
+	if err != nil {
+		return false, err
+	}
+	yamlY, err := yamlSerialize(y)
+	if err != nil {
+		return false, err
+	}
+	return bytes.Equal(yamlX, yamlY), nil
+}
+
+// SecretConfig maps a secret name references to a Kubernetes secret.
+type SecretConfig struct {
+	Name   string                 `yaml:"name"`
+	Config KubernetesSecretConfig `yaml:"config"`
+}
+
+type secretEntry struct {
+	config KubernetesSecretConfig
+	secret Secret
+}
+
+// Manager manages the Kubernetes secret provider.
+type Manager struct {
+	ctx      context.Context
+	cancelFn func()
+	provider *watchProvider
+	config   *WatchSPConfig
+	secrets  map[string]*secretEntry
+	opts     ProviderOptions
+}
+
+// NewManager creates a new secret manager with the provided options.
+func NewManager(ctx context.Context, reg prometheus.Registerer, opts ProviderOptions) Manager {
+	if reg != nil {
+		reg.MustRegister(failedSecretConfigs)
+		reg.MustRegister(secretsTotal)
+	}
+	return Manager{
+		ctx:      ctx,
+		cancelFn: func() {},
+		secrets:  make(map[string]*secretEntry),
+		opts:     opts,
+	}
+}
+
+// ApplyConfig applies the new secrets, diffing each one with the last configuration to apply the
+// relevant update.
+func (m *Manager) ApplyConfig(ctx context.Context, providerConfig *WatchSPConfig, configs []SecretConfig) error {
```

Comment:
question: Do we need a mutex lock here? I see that [discovery](https://github.com/GoogleCloudPlatform/prometheus/blob/TheSpiritXIII/secret-management/discovery/manager.go#L195-L199) uses it, presumably because the config can be reloaded, concurrently, through [various methods](https://github.com/GoogleCloudPlatform/prometheus/blob/TheSpiritXIII/secret-management/cmd/prometheus/main.go#L1016-L1032).